content/blog/2019-09-25-a-verification-backend.md

```diff
@@ -0,0 +1,267 @@
+basic block boundaries. For example, consider the following toy program:
+
+```
+main {
+  a: int = const 2;
+  b: int = const 4;
+  c: int = id a;
+  jmp next;
+next:
+  sum: int = add a c;
+}
+```
+Because, `c` is a copy of `a`, this program would be functionally the same if you replaced the assignment
+to `sum` with `sum: int = add a a`. However, because we are only doing verification on the basic block level,
+we don't know that these programs are equivalent.
+
+Another problem is that this approach to verification relies on the existence of test programs. We are not
+actually analyzing the code of the optimization so if you don't have extensive enough tests, bugs may go by
+unnoticed.
+
+## Evaluation
+To evaluate Shrimp, we implemented [Common sub-expression elimination (CSE)][cse]
+using [Local value numbering (LVN)][lvn] to show that Shrimp is useful in finding
+correctness bugs. We intentionally planted two bugs and found a third bug in the process of testing.
+
+There are some subtleties to a correct implementation of LVN. If you know that the variable
+`sum1` holds the value `a + b`, you have to make sure that `sum1` is not assigned to again before
+you use it. For example, consider the following Bril program:
+```
+sum1: int = add a b;
+sum1: int = id a;
+sum2: int = add a b;
+prod: int = mul sum1 sum2;
+```
+We would like to replace `sum2: int = add a b` with `sum2: int = id sum1` because we
+have already computed the value. However, if we can't do this directly because then `sum2` would
+have the value `a`, not `a + b`. The solution is to rename the first instance of `sum1` to something unique so that we don't lose our reference to the value `a + b`. We can
+then replace `sum2` with a copy from this new variable.
+
+We implemented the faulty version and ran Shrimp. It was able to show that the programs
+are not equivalent and even produced a counter example to prove this.
+With this information, it is easy to walk through the execution of the code
+and discover the source of the bug.
+
+Next we tried extending CSE to deal with associativity.
+It would be nice if the compiler knew that `a + b` is equal to `b + a` so that it could eliminate more
+sub expressions. The most naïve thing to do is sort the arguments for all expressions when you
+compare them so that `a + b` is the same value as `b + a`. However, this by itself is not enough.
+Testing the following example with Shrimp reveals the problem:
+```
+     sub1: int = sub a b;
+     sub2: int = sub b a;
+     prod: int = mul sub1 sub2;
+```
+Shrimp gives us the counter example `a = -8, b = -4`. The problem is that we can't
+sort the arguments for every instruction; `a - b != b - a`. Shrimp helps to reveal
+this problem.
+
+The final bug was actually an unintentional bug that Shrimp helped us find. We made the arguably
+bad decision to give each Bril instruction it's own structure that is a sub-type of a `dest-instr` structure
```

Comment:
```suggestion
bad decision to give each Bril instruction its own structure that is a sub-type of a `dest-instr` structure
```

Reply:
Is `dest-instr` a thing of your own invention? Could use a little intro: "we made a type called `dest-instr` that…"